### ISSUE TITLE
Focus rings on mobile iOS (Safari / Chrome) 

### ISSUE THREAD

**Issue by valeriegohier** [Shopify/dawn#344]
Areas this affects: 

- [ ] [Medias in the media gallery ](https://user-images.githubusercontent.com/53573494/128746651-c4db1ae5-25ce-458c-8f23-a737512aba2f.MP4)
- [ ] The stroke around the item selected in the currency & language switcher in the footer
![Screen Shot 2021-08-09 at 12 55 01 PM](https://user-images.githubusercontent.com/53573494/128744422-2423658e-ba1e-47b1-9364-9c9a5ff25ef1.png)
![Screen Shot 2021-08-09 at 1 26 16 PM](https://user-images.githubusercontent.com/53573494/128747983-f0115eac-131f-4179-b9c2-0fca493c49ff.png)
- [ ] Collapsible tabs on Product Pages ![Screen Shot 2021-08-09 at 1 25 05 PM](https://user-images.githubusercontent.com/53573494/128747835-94e36673-432f-4809-9f05-d0f38a60e964.png)
- [ ] Images on product cards in product grids
- [ ] "view more information" link on Product Page's pick up availability
![Image from iOS](https://user-images.githubusercontent.com/53573494/128748166-f211ca98-245f-4de9-8da0-9455bf269a0d.png)
- [ ] The "X" of the menu navigation
https://user-images.githubusercontent.com/53573494/128749490-ba4b47f5-9113-4778-a9fb-a16a438b9031.MP4

**Comment by valeriegohier** [comment]
iOS 14.7.1

**Comment by LucasLacerdaUX** [comment]
Thanks for providing additional details! I just realized that Safari does not support `focus-visible` at all, even on macOS, so this isn't really a theme issue but a CSS property compatibility limitation. For iOS specifically, every browser is forced to use Safari's rendering engine, meaning you will encounter the same issues on Chrome iOS and any other third-party browser.

For context, let me share a more technical explanation of how our focus approach works:
- We are using two different states to decide whether to show the focus ring:
  - `focus` indicates that the element has received focus through any input type (click, tap, keyboard, etc)
  - `focus-visible` is a new property that allows us to distinguish between elements focused via keyboard interactions (`TAB`) 
- We have a styling rule that sets the focus ring to look like the stroke you're seeing on these elements. This rule is applied to EVERY browser on the `focus` state. 
- **For browsers that support `focus-visible`,** we **disable** this ring on the `focus` state and **restrict** it to `focus-visible`.
- **Browsers that DON'T support `focus-visible`** will **always show** the focus ring regardless of the input type they're using. If we didn't follow this, keyboard users in Safari would not see a focus ring when tabbing through elements.

**TLDR:** `focus-visible`, a property that is not supported in Safari, acts like a progressive enhancement. For compatible browsers, you'll only see these rings when interacting via keyboard. In any different scenario, we need to keep them visible to avoid creating accessibility issues.

Let me know if there are any additional questions or suggestions about this.

**Comment by valeriegohier** [comment]
thank you @LucasLacerdaUX for the information & context. 
But why is that a pattern that I have never encounter in the wild before?

**Comment by melissaperreault** [comment]
**_Accessibility reference_**
> 2.4.7 Focus Visible (Level AA): Any keyboard operable user interface has a mode of operation where the keyboard focus indicator is visible.

Just catching up on the details, and forgot about this reality of incompatibility. I will try to find additional information when/where we took the decision to move forward with this new approach, I know we did surface that at some point.

@svinkle Have you observed other ways to maintain accessible/visible focus for users using keyboard versus a touch device where we wouldn't get the focus when any elements are tapped?

What I hear from @valeriegohier feedback is how might we make the best decisions for merchant's customer to aesthetically not show a bounding box when it is unwanted?

Additionally, can we list our options and also include in the equation the number of potential users experiencing Dawn on Safari perhaps?

I know `focus-visible` [is widely supported except Safari](https://screenshot.click/21-08-jf5z5-z30nj.png) - but if, for instance, Safari is the most popular browser customers will experience Dawn on, we may want to look at all the possibilities offered to us before making our final choice, if we have additional path to resolve the concern.

cc. @jendunnam @Guillaumegranger1 @nicklepine

**Comment by tyleralsbury** [comment]
**TL;DR** - from a purely technical perspective, keeping in mind Dawn's principles, we should stick to the approach that we took.  It's a good approach that future proofs the theme, doesn't add technical debt or JS overhead, and maintains good a11y.  The decision of whether the UX is worth breaking principles is another question, but we need to put some thought into how we make this type of decision to stop ourselves from making breaking our principles a habit.

Part of the question here is whether the principles on which Dawn is built should be bent or broken for the sake of a browser not supporting a particular feature.  This decision was made quite a while back, that we would use `focus-visible` and fall back on `focus` on browsers that don't support it.

If the priorities have changed, we may be able to revisit the decision from the lens of a less principled approach, but there are some downsides to it.

If we use JS, for example, to maybe switch up the styles for Safari, it means more JS, as well as browser-targeted JS.  This can be okay in cases where core functionality i.e. add to cart, checkout, etc... is broken without it, but I would argue that focus ring polish might not be a good enough reason to break this rule unless the UX really is absolutely horrible.

Put another way, does it improve things by enough to justify breaking our principles, and how do we decide whether something is worth breaking principles over?  How do we ensure that we don't start making more and more decisions that break principles?  We need to be careful about that or Dawn as a reference theme will become murky and confusing and it will, over time, become less performant as well.

Another consideration - if we did something like that and then Safari does start to support it, we'll need to remove the workarounds we added since it's now supported.  The more we do this sort of thing, the harder it becomes to keep track of and the more technical debt accrues.

**Comment by tyleralsbury** [comment]
As discussed in our call, I see this as a situation where there are three factors at play, and we can only satisfy two of them.

1. Accessibility
2. UX
3. Technical approach

We came up with some ideas for different combinations of these, sacrificing one each time, so I'll outline them here for further discussion.

## Accessibility + Technical approach

This is what we currently have.  We are sacrificing UX on Safari (both on iOS and on desktop) in order to use the most streamlined technical approach - falling back on the `:focus` selector when `:focus-visible` is not supported on the browser.  However, as a result of using `:focus`, Safari gets focus rings all over the place.  This is a UX issue that will affect a pretty large segment of buyers, which means it's worth looking at.

## Accessibility + UX

This is the best buyer experience, both for abled users and those who are using assistive technologies.  It sacrifices our web native approach slightly in order to incorporate some Javascript to help on Safari.

The idea here would be to have some JS that, ideally only on Safari (if we can detect this), would run whenever the user hits a key and checks what element is focused and adds a `.focused` class onto that element and removes it from whatever element previously had it.

Then when the user taps or clicks on an element, it removes the `.focused` class from the previous element that had it.  In this case, Safari would not fall back onto the `:focus` style and would instead rely on the `.focused` class to add focus rings, thus allowing users who are not using keyboard navigation to not see any focus rings.

@svinkle what do you think of this approach?  It seems to be the one we're leaning towards, given that it's using JS as a progressive enhancement to improve UX without sacrificing any a11y as far as we can tell.

## UX + Technical approach

This approach sacrifices a11y for a small segment of users in order to maintain a very clean technical approach and improve UX for iOS Safari users.  The idea would be to use a media query to at least get rid of focus rings on small screens, thus removing them from iOS.  Note that this wouldn't do anything for the experience for Desktop Safari users, but the larger segment is the iOS users since all browsers on iOS are actually Safari (even Chrome, etc...)

@svinkle I've got a feeling that this is not a preferable approach.  My thought is that there are likely users who use assistive technologies / keyboard navigation on iOS, so we would be alienating those users if we were to take this approach since it would mean no focus rings at all on iOS.

---

I'm leaning towards the UX + Accessibility approach since I don't think the JS will be particularly huge so it won't hurt performance and it can be argued that there's a large enough segment of the market that is using a browser that, for whatever reason, isn't implementing `:focus-visible` so it can be treated as a progressive enhancement for them.

**Comment by tyleralsbury** [comment]
Small follow-up on the JS approach, I've got some (not so well tested) code that should more or less achieve what we're looking for, to give an idea of how much code it is (not much):

```
let currentFocusedElement = null;

window.addEventListener('keydown', (event) => {
  if(event.code.toUpperCase() !== 'TAB') return;

  currentFocusedElement?.classList.remove('focused');

  window.setTimeout(() => {
    currentFocusedElement = document.activeElement;
    currentFocusedElement.classList.add('focused');
  });
});

window.addEventListener('click', (event) => {
  currentFocusedElement.classList.remove('focused');
});
```

Along with this, we'd change all of the references to `:focus` in the CSS to `.focused` instead

**Comment by svinkle** [comment]
@tyleralsbury I'm all for a solution that satisfies our design requirements and accessibility best practices. I'm up for trying your solution but it'd need thorough testing across environments, AT, as well as various templates which feature any complex interaction/focus management.

```js
let currentFocusedElement = null;
```

Would this nullify anything else that uses `document.activeElement`, or the off chance that an app/other third-party script would use `window.currentFocusedElement`?

```js
window.addEventListener('keydown', (event) => {
```

We should instead rely on `keyup` in order to allow [2.5.2: Pointer Cancellation](https://www.w3.org/WAI/WCAG21/Understanding/pointer-cancellation.html) to occur.

**Comment by tyleralsbury** [comment]
I've added an `Effort: Medium` label since the dev work should be somewhat small but the a11y testing will take a bit of time.

**Comment by svinkle** [comment]
> We would probably scope this so it could be block scoped to not use a global variable so it won't step on any toes.

I figured this would be scoped but felt like calling it out. If this was also scoped for Safari only somehow (perhaps user agent sniffing or other feature test) this would help speed up testing.

> I had used `keydown` because that's what seems to actually cause the cursor to move in normal focus

Agreed, this makes sense. Please disregard my earlier comment.
